include/seqan3/io/sam_file/detail/format_sam_base.hpp

```diff
@@ -433,23 +416,24 @@ inline void format_sam_base::read_header(stream_view_type && stream_view,
         }
     };
 
-    auto skip_tag_name_and_colon = [&it] ()
+    auto skip_until_predicate = [&it] (auto const & predicate)
     {
-        while (!is_char<':'>(*it))
+        while (!predicate(*it))
             ++it;
-        ++it;
     };
 
     auto parse_tag_value = [&] (auto & value) // helper function to parse the next tag value
     {
-        skip_tag_name_and_colon();
+        skip_until_predicate(is_char<':'>);
+        ++it;                                                           // skip :
         take_until_predicate(is_char<'\t'> || is_char<'\n'>);
         read_field(string_buffer, value);
     };
 
-    auto parse_until_newline = [&] (auto & value)
+    auto parse_until_tab_or_newline = [&] (auto & value)
     {
-        take_until_predicate(is_char<'\n'>);
+        take_until_predicate(is_char<'\t'> || is_char<'\n'>);
+        string_buffer.push_back('\t');
```

Comment:
No, without the push_back() we would get unseperated tags:
```bash
header.read_groups[0]
    Which is: ("U0a_A2_L1", "PL:illuminaPU:1LB:1SM:NA12878")
  (std::pair<std::string, std::string>{"U0a_A2_L1", "PL:illumina\tPU:1\tLB:1\tSM:NA12878"})
    Which is: ("U0a_A2_L1", "PL:illumina\tPU:1\tLB:1\tSM:NA12878")
```

Reply:
Ah.

If I see it correctly, you could just do:

```
take_until_predicate(is_char<'\t'> || is_char<'\n'>);
if (is_char<'\t'>(*it))
    string_buffer.push_back('\t');
read_field(string_buffer, value);
```

After `take_until_predicate` the iterator should point to a tab or newline, and you only need a tab in your buffer if you are not at the end.

Then you can also get rid of `pop_back` further down.

Reply:
That would be nice, but this doesn't work eather. When we take the example again, the sam file line looks like:
`@RG	ID:U0a_A2_L1	PL:illumina	PU:1	LB:1	SM:NA12878`
but it could also look like
`@RG	PL:illumina	PU:1	LB:1	SM:NA12878	ID:U0a_A2_L1`
The ID is not a part of the second part of the `read_groups` tuple, thus the `\n` would not be at the end of the elements of this second part and we would never avoid the `if`.

->
```bash
.../seqan3/test/unit/io/sam_file/sam_file_format_test_template.hpp:183: Failure
Expected equality of these values:
  header.read_groups[0]
    Which is: ("U0a_A2_L1", "PL:illumina\tPU:1\tLB:1\tSM:NA12878\t")
  (std::pair<std::string, std::string>{"U0a_A2_L1", "PL:illumina\tPU:1\tLB:1\tSM:NA12878"})
    Which is: ("U0a_A2_L1", "PL:illumina\tPU:1\tLB:1\tSM:NA12878")
```

Reply:
Thanks for the explanation :)
Can you add this in a condensed form as comment to the lambda?

I'll take another look sometime later, but it doesn't seem like it can be simplified.

Reply:
I managed to find a way